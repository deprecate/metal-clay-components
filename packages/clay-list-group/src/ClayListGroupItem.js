import 'clay-dropdown';
import 'clay-icon';
import 'clay-label';
import 'clay-link';
import 'clay-sticker';

import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayListGroupItem.soy.js';

/**
 * Metal ClayListGroupItem component.
 */
class ClayListGroupItem extends Component {
	/**
	 * Continues the propagation of the checkbox changed event
	 * @param {!Event} event
	 * @private
	 */
	handleItemCheckboxClick_(event) {
		this.emit('itemToggled', event);
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayListGroupItem.STATE = {
	/**
	 * List of items to display in the actions menu.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {!array}
	 * @default undefined
	 */
	actionItems: Config.arrayOf(
		Config.shapeOf({
			disabled: Config.bool().value(false),
			icon: Config.string(),
			label: Config.string().required(),
			quickAction: Config.bool(),
			separator: Config.bool().value(false),
			url: Config.string().required(),
		})
	),

	/**
	 * Name of the content renderer to use template variants.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	contentRenderer: Config.string(),

	/**
	 * Description of the item
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	description: Config.string(),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Icon of the list item.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	icon: Config.string(),

	/**
	 * Shape of the icon of the list item.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default rounded
	 */
	iconShape: Config.oneOf(['circle', 'rounded']).value('rounded'),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Label of the list item.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	label: Config.string(),

	/**
	 * Style of the label of the list item.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default secondary
	 */
	labelStyle: Config.oneOf([
		'danger',
		'info',
		'secondary',
		'success',
		'warning',
	]).value('secondary'),

	/**
	 * Flag to indicate if the list item is selectable.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?bool|undefined}
	 * @default false
	 */
	selectable: Config.bool().value(false),

	/**
	 * Flag to indicate if the list item is selected.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?bool|undefined}
	 * @default false
	 */
	selected: Config.bool().value(false),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	spritemap: Config.string(),

	/**
	 * Title of the list item.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	title: Config.string(),

	/**
	 * Url of the item.
	 * @instance
	 * @memberof ClayListGroupItem
	 * @type {?string|undefined}
	 * @default undefined
	 */
	url: Config.string(),
};

defineWebComponent('clay-list-item', ClayListGroupItem);

Soy.register(ClayListGroupItem, templates);

export {ClayListGroupItem};
export default ClayListGroupItem;
