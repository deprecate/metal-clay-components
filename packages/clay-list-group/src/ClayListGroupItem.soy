{namespace ClayListGroupItem}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param title: string}
	{@param? actionItems: list<[
			disabled: bool,
			icon: string,
			label: string,
			quickAction: bool,
			separator: bool,
			url: string
	]>}
	{@param? contentRenderer: string}
	{@param? description: string}
	{@param? elementClasses: string}
	{@param? handleItemCheckboxClick_: any}
	{@param? icon: string}
	{@param? iconShape: string}
	{@param? id: string}
	{@param? label: string}
	{@param? labelStyle: string}
	{@param? selectable: bool}
	{@param? selected: bool}
	{@param? spritemap: string}
	{@param? url: string}

	{let $listItemAttributes kind="attributes"}
    class="list-group-item list-group-item-flex
  		{if $elementClasses}
	  		{sp}{$elementClasses}
		  {/if}

  		{if $selected}
	  		{sp}active
		  {/if}
    "

    {if $id}
      id="{$id}"
    {/if}
	{/let}

	<li {$listItemAttributes}>
		{let $isSimple: not $description and not $selectable and not $icon and not $actionItems and not $label/}

		{if $isSimple}
			{delcall ClayListGroupItem.Content variant="$contentRenderer"}
				{param title: $title /}
				{param url: $url /}
			{/delcall}
		{else}
			{if $selectable}
				<div class="flex-col">
					{call ClayCheckbox.render}
						{param checked: $selected /}
						{param events: ['change': $handleItemCheckboxClick_] /}
						{param hideLabel: true /}
						{param label: 'select' /}
					{/call}
				</div>
			{/if}

			{if $icon and $spritemap}
				{delcall ClayListGroupItem.Icon variant="$contentRenderer"}
					{param icon: $icon /}
					{param iconShape: $iconShape /}
					{param spritemap: $spritemap /}
				{/delcall}
			{/if}

			{delcall ClayListGroupItem.Content variant="$contentRenderer"}
				{param description: $description /}
				{param label: $label /}
				{param labelStyle: $labelStyle /}
				{param spritemap: $spritemap /}
				{param title: $title /}
				{param url: $url /}
			{/delcall}

			{if $actionItems and $spritemap}
        <div class="flex-col">
					{call .quickMenu}
						{param items: $actionItems /}
						{param spritemap: $spritemap /}
					{/call}

					{call ClayActionsDropdown.render}
						{param items: $actionItems /}
						{param spritemap: $spritemap /}
					{/call}
				</div>
			{/if}
		{/if}
	</li>
{/template}

/**
 * This renders the quick actions menu
 */
{template .quickMenu}
	{@param items: list<[
		disabled: bool,
		icon: string,
		label: string,
		quickAction: bool,
		separator: bool,
		url: string
	]>}
	{@param spritemap: string}

	<div class="quick-action-menu">
		{foreach $item in $items}
			{if $item.quickAction and $item.icon}
				{call ClayLink.render}
					{param elementClasses: 'quick-action-item' /}
					{param label kind="html"}
						{call ClayIcon.render}
							{param spritemap: $spritemap /}
							{param symbol: $item.icon /}
						{/call}
					{/param}
				{/call}
			{/if}
		{/foreach}
	</div>
{/template}

/**
 * This renders the content of the item.
 */
{deltemplate ClayListGroupItem.Content}
	{@param title: string}
	{@param? description: string}
	{@param? label: string}
	{@param? labelStyle: string}
	{@param? spritemap: string}
	{@param? url: string}

	<div class="flex-col flex-col-expand">
		<h4 class="list-group-title text-truncate">
			{if $url}
				{call ClayLink.render}
					{param href: $url /}
					{param label: $title /}
				{/call}
			{else}
				{$title}
			{/if}
		</h4>

		{if $description}
			<p class="list-group-subtitle text-truncate">{$description}</p>
		{/if}

		{if $label}
			<div class="list-group-detail">
				{call ClayLabel.render}
					{param label: $label /}
					{param style: $labelStyle /}
				{/call}
			</div>
		{/if}
	</div>
{/deltemplate}

/**
 * This renders the icon of the item.
 */
{deltemplate ClayListGroupItem.Icon}
	{@param icon: string}
	{@param spritemap: string}
	{@param? iconShape: string}

	<div class="flex-col">
		{call ClaySticker.render}
			{param icon: $icon /}
			{param shape: $iconShape /}
			{param spritemap: $spritemap /}
		{/call}
	</div>
{/deltemplate}

/**
 * This renders a simple item.
 */
{deltemplate ClayListGroupItem.Simple}
	{@param title: string}
	{@param? url: string}

	{if $url}
		{call ClayLink.render}
			{param href: $url /}
			{param label: $title /}
		{/call}
	{else}
		{$title}
	{/if}
{/deltemplate}