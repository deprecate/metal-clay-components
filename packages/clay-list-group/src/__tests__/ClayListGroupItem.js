import ClayListGroupItem from '../ClayListGroupItem';

const spritemap = '../node_modules/lexicon-ux/build/images/icons/icons.svg';

let component;

let actionItems = [
	{
		label: 'Option 1',
		url: '#1',
	},
	{
		label: 'Option 2',
		separator: true,
		url: '#2',
	},
	{
		label: 'Option 3',
		url: '#3',
	},
];

let actionItemsWithQuickItems = [
	{
		icon: 'trash',
		label: 'Option 1',
		quickAction: true,
		url: '#1',
	},
	{
		icon: 'download',
		label: 'Option 2',
		quickAction: true,
		separator: true,
		url: '#2',
	},
	{
		icon: 'info-circle-open',
		label: 'Option 3',
		quickAction: true,
		url: '#3',
	},
];

describe('ClayListGroupItem', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayListGroupItem({
			title: 'My Item',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with classes', () => {
		component = new ClayListGroupItem({
			elementClasses: 'my-custom-class',
			title: 'My Item',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with id', () => {
		component = new ClayListGroupItem({
			id: 'myId',
			title: 'My Item',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with link', () => {
		component = new ClayListGroupItem({
			title: 'Item 1',
			url: '#1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with description', () => {
		component = new ClayListGroupItem({
			description: 'Description 1',
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with description and url', () => {
		component = new ClayListGroupItem({
			description: 'Description 1',
			title: 'Item 1',
			url: '#1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selectable ClayListGroupItem', () => {
		component = new ClayListGroupItem({
			description: 'Description 1',
			selectable: true,
			title: 'Item 1',
			url: '#1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selected ClayListGroupItem', () => {
		component = new ClayListGroupItem({
			description: 'Description 1',
			selectable: true,
			selected: true,
			title: 'Item 1',
			url: '#1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with icon', () => {
		component = new ClayListGroupItem({
			icon: 'folder',
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with circle icons', () => {
		component = new ClayListGroupItem({
			icon: 'folder',
			iconShape: 'circle',
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with one label', () => {
		component = new ClayListGroupItem({
			labels: [
				{
					label: 'Status',
				},
			],
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with two labels', () => {
		component = new ClayListGroupItem({
			labels: [
				{
					label: 'Aproved',
				},
				{
					label: 'Pending',
				},
			],
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with labels with style', () => {
		component = new ClayListGroupItem({
			labels: [
				{
					label: 'Aproved',
					style: 'info',
				},
				{
					label: 'Pending',
					style: 'warning',
				},
			],
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with action menu', () => {
		component = new ClayListGroupItem({
			actionItems: actionItems,
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem with quick action menu', () => {
		component = new ClayListGroupItem({
			actionItems: actionItemsWithQuickItems,
			spritemap: spritemap,
			title: 'Item 1',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayListGroupItem and emit an event on item toggled', () => {
		const spy = jest.fn();

		component = new ClayListGroupItem({
			events: {itemToggled: spy},
			selectable: true,
			title: 'Item 1',
		});

		component.element.querySelector('input[type="checkbox"]').click();

		expect(spy).toHaveBeenCalled();
	});

	it('should fail when no title is passed', function() {
		expect(() => {
			component = new ClayListGroupItem();
		}).toThrow();
	});
});
