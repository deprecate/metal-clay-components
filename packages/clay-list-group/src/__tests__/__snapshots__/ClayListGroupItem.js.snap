// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayListGroupItem should render a ClayListGroupItem with action menu 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
  <div class="flex-col">
    <div class="quick-action-menu"></div>
    <div class="dropdown dropdown-action">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#ellipsis-v"></use>
        </svg>
      </button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class=" dropdown-item" href="#1">Option 1</a>
          </li>
          <li>
            <a class=" dropdown-item" href="#2">Option 2</a>
          </li>
          <li aria-hidden="true" class="dropdown-divider" role="presentation"></li>
          <li>
            <a class=" dropdown-item" href="#3">Option 3</a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a ClayListGroupItem with circle icons 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col">
    <span class="sticker sticker-primary rounded-circle">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#folder"></use>
      </svg>
    </span>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a ClayListGroupItem with classes 1`] = `<li class="list-group-item list-group-item-flex my-custom-class">My Item</li>`;

exports[`ClayListGroupItem should render a ClayListGroupItem with description 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a ClayListGroupItem with description and url 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">
      <a href="#1">Item 1</a>
    </h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a ClayListGroupItem with icon 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col">
    <span class="sticker sticker-primary">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-folder">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#folder"></use>
      </svg>
    </span>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a ClayListGroupItem with id 1`] = `<li class="list-group-item list-group-item-flex" id="myId">My Item</li>`;

exports[`ClayListGroupItem should render a ClayListGroupItem with label 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
    <div class="list-group-detail">
      <span class="label label-secondary">Status</span>
    </div>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a ClayListGroupItem with link 1`] = `
<li class="list-group-item list-group-item-flex">
  <a href="#1">Item 1</a>
</li>
`;

exports[`ClayListGroupItem should render a ClayListGroupItem with quick action menu 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">Item 1</h4>
  </div>
  <div class="flex-col">
    <div class="quick-action-menu">
      <a class=" quick-action-item">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-trash">
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#trash"></use>
        </svg>
      </a>
      <a class=" quick-action-item">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-download">
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#download"></use>
        </svg>
      </a>
      <a class=" quick-action-item">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle-open">
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle-open"></use>
        </svg>
      </a>
    </div>
    <div class="dropdown dropdown-action">
      <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-unstyled">
        <svg aria-hidden="true" class="lexicon-icon lexicon-icon-ellipsis-v">
          <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#ellipsis-v"></use>
        </svg>
      </button>
      <div class="dropdown-menu">
        <ul class="list-unstyled">
          <li>
            <a class=" dropdown-item" href="#1">Option 1</a>
          </li>
          <li>
            <a class=" dropdown-item" href="#2">Option 2</a>
          </li>
          <li aria-hidden="true" class="dropdown-divider" role="presentation"></li>
          <li>
            <a class=" dropdown-item" href="#3">Option 3</a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a selectable ClayListGroupItem 1`] = `
<li class="list-group-item list-group-item-flex">
  <div class="flex-col">
    <div class="custom-control custom-checkbox">
      <label>
        <input class="custom-control-input" ref="input" type="checkbox">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description sr-only">select</span>
      </label>
    </div>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">
      <a href="#1">Item 1</a>
    </h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListGroupItem should render a selected ClayListGroupItem 1`] = `
<li class="list-group-item list-group-item-flex active">
  <div class="flex-col">
    <div class="custom-control custom-checkbox">
      <label>
        <input checked="" class="custom-control-input" ref="input" type="checkbox">
        <span class="custom-control-indicator"></span>
        <span class="custom-control-description sr-only">select</span>
      </label>
    </div>
  </div>
  <div class="flex-col flex-col-expand">
    <h4 class="list-group-title text-truncate">
      <a href="#1">Item 1</a>
    </h4>
    <p class="list-group-subtitle text-truncate">Description 1</p>
  </div>
</li>
`;

exports[`ClayListGroupItem should render the default markup 1`] = `<li class="list-group-item list-group-item-flex">My Item</li>`;
